lib/AST/GenericSignatureBuilder.cpp

```diff
@@ -6905,34 +6907,121 @@ void GenericSignatureBuilder::checkSameTypeConstraints(
   // connect all of the components, or else we wouldn't have an equivalence
   // class.
   if (intercomponentEdges.size() > numComponents - 1) {
-    std::vector<bool> connected(numComponents, false);
-    const auto &firstEdge = intercomponentEdges.front();
-    for (const auto &edge : intercomponentEdges) {
-      // If both the source and target are already connected, this edge is
-      // not part of the spanning tree.
-      if (connected[edge.source] && connected[edge.target]) {
-        if (edge.constraint.source->shouldDiagnoseRedundancy(true) &&
-            firstEdge.constraint.source->shouldDiagnoseRedundancy(false)) {
-          Diags.diagnose(edge.constraint.source->getLoc(),
-                         diag::redundant_same_type_constraint,
-                         edge.constraint.getSubjectDependentType(
-                                                          genericParams),
-                         edge.constraint.value);
-
-          Diags.diagnose(firstEdge.constraint.source->getLoc(),
-                         diag::previous_same_type_constraint,
-                         firstEdge.constraint.source->classifyDiagKind(),
-                         firstEdge.constraint.getSubjectDependentType(
-                                                          genericParams),
-                         firstEdge.constraint.value);
+    auto &sourceMgr = getASTContext().SourceMgr;
+
+    // First let's order all of the intercomponent edges
+    // as written in source, this helps us to diagnose
+    // all of the duplicate constraints in correct order.
+    std::vector<unsigned> sourceOrderedEdges;
+    for (unsigned i : indices(intercomponentEdges))
+      sourceOrderedEdges.push_back(i);
+
+    std::sort(sourceOrderedEdges.begin(), sourceOrderedEdges.end(),
```

Comment:
Is there any way we can avoid sorting in the common case where there aren't any diagnostics to emit? The `GenericSignatureBuilder` is very much a hot path in the type checker.

Reply:
Sure, I can use `array_pod_sort` there, but it seems like this is not a common path because it's been determined that we have more inter-component edges than components and it has to be diagnosed?...

Reply:
Hmm... I see what you mean, okay!

Reply:
Moved to `llvm::array_pod_sort`, thanks for a suggestion!